lib/puppetserver/ca/action/generate.rb

```diff
@@ -202,13 +202,22 @@ def generate_certs(certnames, alt_names, settings, digest)
             key, csr = generate_key_csr(certname, settings, digest, current_alt_names)
             next false unless csr
             next false unless ca.submit_certificate_request(certname, csr)
-            next false unless ca.sign_certs([certname])
+
+            # Check if the CA autosigned the cert
             if result = ca.get_certificate(certname)
+              @logger.inform "Certificate for #{certname} was autosigned."
               next false unless save_file(result.body, certname, settings[:certdir], "Certificate")
               next false unless save_keys(certname, settings, key)
               true
             else
-              false
+              next false unless ca.sign_certs([certname])
+              if result = ca.get_certificate(certname)
+                next false unless save_file(result.body, certname, settings[:certdir], "Certificate")
```

Comment:
If signing fails should we save the csr?

Reply:
If we do, does this need to also try to load it before generating it? And if we save the CSR we probably also need to save the keys that go with it.

Reply:
I'm into saving the keys regardless. I'm indifferent to loading them. I could see reading bits from disk trying to pick up a failed generation at the right spot being a little magical and error prone.

The use case that I run into is that I do a `puppetserver ca generate --certname foo --subject-alt-name bar` and it submits a CSR for said certname but fails with a "update your server conf and restart your server to allow signing SANS". I do that, but then I have a CSR waiting on the server (that I can now sign) but I don't have the keys for it even if I do.

I'd be happy if there's an error the cli just said, "Help, here are the artifacts I created, you're going to have to resolve this on your own".